sdcflows/interfaces/bspline.py

```diff
@@ -474,12 +474,13 @@ def _fix_topup_fieldcoeff(in_coeff, fmap_ref, refpe_reversed=False, out_file=Non
     refnii = nb.load(fmap_ref)
 
     coeff_shape = np.array(coeffnii.shape[:3])
+    factors = np.array(coeffnii.header.get_zooms()[:3])
     ref_shape = np.array(refnii.shape[:3])
-    factors = coeffnii.header.get_zooms()[:3]
-    if not np.all(coeff_shape == ref_shape // factors + 3):
+    exp_shape = ref_shape // factors + 3 if np.all(factors > 1.0) else ref_shape
```

Comment:
The title says "larger than 2mm", this says `np.all(factors > 1)`. I'm having trouble squaring either choice with the docstring:

```
    TOPUP-generated "fieldcoeff" files are just B-Spline fields, where the shape
    of the field is fixated to be a decimated grid of the original image by an
    integer factor and added 3 pixels on each dimension.
```

Given that this seems to correspond to choice of `--warpres` passed to `topup` (apologies if I'm misreading the conversation), I wonder if we should take that value as a parameter here. Making the calculations explicit would be much clearer, IMO.

Reply:
The title is definitely inaccurate - it's about the integer factors that determine how far apart control knots are w.r.t. the original image's resolution.

As per the documentation:

> --warpres
> 
> Specifies the resolution (in mm) of the warps. Specifying e.g. --warpres=10 means that an isotropic knot-spacing of 10mm is used for all levels of --subsamp. If instead specifying e.g. --warpres=10,5 indicates that a (n isotropic) knot-spacing of 10mm should be used for the first level of --subsamp followed by 5mm for the second level. A warp resolution of e.g. 10mm does not imply that 10mm is the highest accuracy that can be obtained for registration of any given structure. It is rather related to how fast the field can change as on goes from one point to the next in the field. The warps/field are implemented as cubic/quadratic B-splines with a knot-spacing that has to be an integer multiple of the voxel-size of the --imain images. If a value is specified for --warpres that is not and integer multiple the value will be rounded down. So if for example the voxel-size of the --imain image is 3x3x4mm and --warpes is specified as --warpres=10 the actual resolution (knot-spacing) of the warps will be 9x9x8mm.
> 
> When running topup it is advantageous to go to a quite high warp-resolution, typically all the way to the voxel size (i.e. one spline per voxel). To achieve an actual knot-spacing of 1 voxel in all directions one should specify a knot-spacing smaller than two voxels in the direction with the smallest voxel size. If one e.g. has isotropic voxels with size 2.5x2.5x3.5mm, specifying --warpres=4 (4<2*2.5) will give an actual knot-spacing of 2.5x2.5x3.5mm (i.e. 1x1x1voxels).
> 
> Increasing the resolution means that topup will need more execution time and more memory. Most of the time and memory is spent on estimating and representing the Hessian. It may therefore be advantageous to swith to an alternative minimization method (i.e. Scaled Conjugate Gradient, specified by --minmet=1) for the final high resolution steps.

The resulting factors are then encoded as the zooms in the coefficients NIfTI file.

> Given that this seems to correspond to the choice of ``--warpres`` passed to ``topup`` (apologies if I'm misreading the conversation), I wonder if we should take that value as a parameter here. Making the calculations explicit would be much clearer, IMO.

So the decision `topup` makes internally is encoded on the size and spacings of the coefficients NIfTI file we get at the output, and as @Lestropie pointed out in the linked issue, `topup` is indeed running.

Then, as @madisoth reported, when these factors are 1x1x1 (and I'm not sure it is possible to have anisotropic factors when one of them is 1, as per `--warpres` documentation), we have that the coefficients matrix is not added 3 extra pixels to better cover the FoV with fully supported B-Splines.

Reply:
I see. I think my brain was not in a place to fully parse that docstring yesterday. This makes much more sense now.

I think we probably can achieve anisotropic factors with some 1 based on the docstring. For example, `--warpres 6` with voxels `(2.5, 2.5, 3.5)` should result in `(2, 2, 1)`. Note that they say that it should be `<2*min(zoom)` to achieve isotropic 1, not `<2*max(zoom)`.

So I think what you want is either:

```suggestion
    exp_shape = ref_shape // factors + 3 if np.any(factors > 1) else ref_shape
```

Or possibly

```suggestion
    exp_shape = ref_shape // factors + 3 * (factors > 1)
```

Depending on whether they pad in case of any scaling or only on scaled directions.

In both cases I've switched to a integer 1 for the sake of making it clearer that it's an integer comparison. You might consider making `factors = np.uint8(coeffnii.header.get_zooms()[:3])` to be even more explicit.

Reply:
As per this bit of the topup documentation:

> To achieve an actual knot-spacing of 1 voxel in all directions one should specify a knot-spacing smaller than two voxels in the direction with the smallest voxel size. If one e.g. has isotropic voxels with size 2.5x2.5x3.5mm, specifying --warpres=4 (4<2*2.5) will give an actual knot-spacing of 2.5x2.5x3.5mm (i.e. 1x1x1voxels).

I'm unsure the factors will not be all ones when the smallest zoom is more than half warpres.

Reply:
That said, your second suggestion should work either way